nsxt/resource_nsxt_manager_cluster.go

```diff
@@ -0,0 +1,441 @@
+				ID:        id,
+				IPAddress: ip,
+				Fqdn:      fqdn,
+				Status:    status,
+			}
+			nodes = append(nodes, clusterNode)
+		}
+
+	}
+	d.Set("revision", clusterConfig.Revision)
+	setNodesInSchema(d, nodes)
+	return nil
+}
+
+func getIPFromNodeInfo(node nsxModel.ClusterNodeInfo, isIPv4 bool) string {
+	// After join node into cluster, apiListen address may take some time to become available
+	// if it's not available, we retrieve node ip from nodeEntityInfo
+	var ip *string
+	nodeEntities := node.Entities
+	for _, entity := range nodeEntities {
+		if *entity.Port == 443 {
+			if isIPv4 {
+				ip = entity.IpAddress
+			} else {
+				ip = entity.Ipv6Address
+			}
+			break
+		}
+	}
+	return *ip
+}
+
+func resourceNsxtManagerClusterUpdate(d *schema.ResourceData, m interface{}) error {
+	connector := getPolicyConnector(m)
+	client := nsx.NewClusterClient(connector)
+
+	clusterID, certSha256Thumbprint, hostIP, err := getClusterInfoFromHostNode(d, m)
+	if err != nil {
+		return err
+	}
+	oldNodes, newNodes := d.GetChange("node")
+	oldNodesIPs := getClusterNodesIPs(oldNodes)
+	newNodesIPs := getClusterNodesIPs(newNodes)
+	for _, node := range oldNodes.([]interface{}) {
+		nodeMap := node.(map[string]interface{})
+		ip := nodeMap["ip_address"].(string)
+		if !slices.Contains(newNodesIPs, ip) {
+			id := nodeMap["id"].(string)
+			force := "true"
+			gracefulShutdown := "true"
+			ignoreRepositoryIPCheckParam := "false"
+			_, err := client.Removenode(id, &force, &gracefulShutdown, &ignoreRepositoryIPCheckParam)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	for _, node := range newNodes.([]interface{}) {
+		nodeMap := node.(map[string]interface{})
+		ip := nodeMap["ip_address"].(string)
+		if !slices.Contains(oldNodesIPs, ip) {
+			userName := nodeMap["username"].(string)
+			password := nodeMap["password"].(string)
+			ip := nodeMap["ip_address"].(string)
+			nodeObj := NsxClusterNode{
+				IPAddress: ip,
+				UserName:  userName,
+				Password:  password,
+			}
+			err = joinNodeToCluster(clusterID, certSha256Thumbprint, nodeObj, hostIP, d, m)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return resourceNsxtManagerClusterRead(d, m)
+}
+
+func getClusterNodesIPs(nodes interface{}) []string {
+	var ips []string
+	for _, node := range nodes.([]interface{}) {
+		nodeMap := node.(map[string]interface{})
+		ips = append(ips, nodeMap["ip_address"].(string))
+	}
+	return ips
+}
+
+func resourceNsxtManagerClusterDelete(d *schema.ResourceData, m interface{}) error {
+	connector := getPolicyConnector(m)
+	client := nsx.NewClusterClient(connector)
+	nodes := getClusterNodesFromSchema(d)
+	force := "true"
+	gracefulShutdown := "true"
+	ignoreRepositoryIPCheckParam := "false"
+	for _, node := range nodes {
+		guestNodeID := node.ID
+		_, err := client.Removenode(guestNodeID, &force, &gracefulShutdown, &ignoreRepositoryIPCheckParam)
+		if err != nil {
+			return fmt.Errorf("Failed to remove node %s from cluster: %s", guestNodeID, err)
```

Comment:
Similar to above, lets use `handleDeleteError` here to provide more details to the user